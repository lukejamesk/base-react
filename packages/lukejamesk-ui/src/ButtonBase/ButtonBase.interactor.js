import { createInteractor } from '@lukejamesk/test-utils'

const buttonElements = (wrapper) => ({
  button() {
    return wrapper.find('[type="button"]')
  },
})

const buttonActions = (elements) => ({
  click() {
    elements.button().simulate('click')
    return this
  },
})

const ButtonBaseInteractor = createInteractor(buttonElements, buttonActions, undefined)

export default ButtonBaseInteractor
