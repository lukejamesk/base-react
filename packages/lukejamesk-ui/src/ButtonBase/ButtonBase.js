import React from 'react'
import PropTypes from 'prop-types'

const ButtonBase = ({ children, className, element: Element, ...props }) => {
  return (
    <Element type="button" className={className} {...props} tabIndex="0">
      {children}
    </Element>
  )
}

ButtonBase.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  element: PropTypes.string,
}

ButtonBase.defaultProps = {
  element: 'button',
}

export default ButtonBase
