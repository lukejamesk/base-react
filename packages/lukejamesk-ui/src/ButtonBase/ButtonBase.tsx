import React from 'react'
import PropTypes from 'prop-types'

export interface ButtonBaseProps extends React.HTMLProps<HTMLButtonElement> {
  children: React.ReactNode
  className?: string
  element?: string | React.FunctionComponent<React.HTMLProps<HTMLButtonElement>>
}

const ButtonBase: React.FunctionComponent<ButtonBaseProps> = ({
  children,
  className,
  element: Element = 'button',
  ...props
}) => {
  return (
    <Element type="button" className={className} tabIndex={0} {...props}>
      {children}
    </Element>
  )
}

ButtonBase.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  element: PropTypes.string,
}

export default ButtonBase
