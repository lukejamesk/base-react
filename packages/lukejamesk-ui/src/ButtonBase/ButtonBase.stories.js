import React from 'react'
import { text } from '@storybook/addon-knobs'
import ButtonBase from './ButtonBase'

export default {
  title: 'Core|ButtonBase',
}

export const Default = () => <ButtonBase>{text('text', 'My Button')}</ButtonBase>
