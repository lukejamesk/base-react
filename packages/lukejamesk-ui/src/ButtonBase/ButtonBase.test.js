import React from 'react'
import { render } from 'enzyme'
import ButtonBase from './ButtonBase'
import ButtonBaseInteractor from './ButtonBase.interactor'

describe('<ButtonBase />', () => {
  it('Should render', () => {
    const wrapper = render(<ButtonBase>My content for button</ButtonBase>)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should allow a to use an alternate element', () => {
    const clickSpy = jest.fn()
    const interactor = ButtonBaseInteractor(
      <ButtonBase element="span" onClick={clickSpy}>
        My content for button
      </ButtonBase>,
    )

    expect(interactor.elements.button().type()).toEqual('span')
    interactor.actions.click()
    expect(clickSpy.mock.calls.length).toEqual(1)
  })

  describe('Actions:', () => {
    let clickSpy
    let interactor
    beforeEach(() => {
      clickSpy = jest.fn()
      interactor = ButtonBaseInteractor(<ButtonBase onClick={clickSpy}>My content for button</ButtonBase>)
    })
    it('should fire onClick callback', () => {
      interactor.actions.click()
      expect(clickSpy.mock.calls.length).toEqual(1)
    })
  })
})
