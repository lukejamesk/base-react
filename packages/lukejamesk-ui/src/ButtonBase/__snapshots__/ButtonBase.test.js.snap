// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonBase /> Should render 1`] = `
<button
  tabindex="0"
  type="button"
>
  My content for button
</button>
`;
